src/ringmesh/io/geomodel_builder_resqml.cpp

```diff
@@ -0,0 +1,994 @@
+    };
+
+    GeoModelBuilderRESQMLImpl::GeoModelBuilderRESQMLImpl(
+        GeoModelBuilderRESQML& builder,
+        GeoModel3D& geomodel,
+        GeoModelAccess< 3 >& geomodel_access )
+        : builder_( builder ),
+          geomodel_( geomodel ),
+          geomodel_access_( geomodel_access )
+    {
+    }
+
+    void GeoModelBuilderRESQMLImpl::deserialize( EpcDocument& pck )
+    {
+        std::string resqmlResult = pck.deserialize();
+        if( !resqmlResult.empty() )
+        {
+            Logger::err( "", "RESQML2 document deserialization failed..." );
+            return;
+        }
+
+        Logger::out( "", "EpcDocument name ", pck.getName(), " in ",
+            ( pck.getStorageDirectory().empty() ? "working directory."
+                                                : pck.getStorageDirectory() ) );
+
+        unsigned int hdfProxyCount = pck.getHdfProxyCount();
+        Logger::out( "", "There are ", pck.getHdfProxyCount(),
+            " hdf files associated to this epc document." );
+        for( unsigned int hdfProxyIndex = 0; hdfProxyIndex < hdfProxyCount;
+             ++hdfProxyIndex )
+        {
+            Logger::out( "", "Hdf file relative path : ",
+                pck.getHdfProxy( hdfProxyIndex )->getRelativePath() );
+        }
+        for( size_t warningIndex = 0; warningIndex < pck.getWarnings().size();
+             ++warningIndex )
+        {
+            Logger::out( "", "Warning #", warningIndex, " : ",
+                pck.getWarnings()[warningIndex] );
+        }
+    }
+
+    namespace
+    {
+        RELATION get_contact_mode(
+            gsoap_resqml2_0_1::resqml2__ContactMode mode )
+        {
+            switch( mode )
+            {
+            case gsoap_resqml2_0_1::resqml2__ContactMode__baselap:
+                return RELATION::BASELAP;
+            case gsoap_resqml2_0_1::resqml2__ContactMode__erosion:
+                return RELATION::ERODED;
+            case gsoap_resqml2_0_1::resqml2__ContactMode__proportional:
+                return RELATION::CONFORMABLE;
+            default:
+                ringmesh_assert_not_reached;
+                return RELATION::CONFORMABLE;
+            }
+        }
+    } // namespace
+
+    gmge_id GeoModelBuilderRESQMLImpl::find_layer(
+        const Region3D& region ) const
+    {
+        std::set< const Interface3D* > horizons;
+        for( auto b : range( region.nb_boundaries() ) )
+        {
+            const Surface3D& surface = region.boundary( b );
+            const Interface3D& interface = static_cast< const Interface3D& >(
+                surface.parent( Interface3D::type_name_static() ) );
+            if( GMGE::is_stratigraphic_limit( interface.geological_feature() ) )
+            {
+                horizons.insert( &interface );
+            }
+        }
+
+        if( horizons.size() >= 1 && horizons.size() <= 2 )
+        {
+            // TODO: the following detection is not working if the region
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS1135)